eth/beacon/db/chain.py

```diff
@@ -0,0 +1,450 @@
+
+        Raises BlockNotFound if there's no block with the given number in the
+        canonical chain.
+
+        NOTE: copied from ``eth.db.header.HeaderDB.get_canonical_block_hash``,
+        and changed "header" to "block".
+        """
+        return self._get_canonical_block_hash(self.db, slot)
+
+    @staticmethod
+    def _get_canonical_block_hash(db: BaseDB, slot: int) -> Hash32:
+        """
+        NOTE: copied from ``eth.db.header.HeaderDB._get_canonical_block_hash``,
+        and changed "header" to "block".
+        """
+        validate_block_number(slot, title="Slot")  # FIXME: Create validate_slot
+        slot_to_hash_key = SchemaV1.make_block_slot_to_hash_lookup_key(slot)
+        try:
+            encoded_key = db[slot_to_hash_key]
+        except KeyError:
+            raise BlockNotFound(
+                "No canonical block for block slot #{0}".format(slot)
+            )
+        else:
+            return rlp.decode(encoded_key, sedes=rlp.sedes.binary)
+
+    def get_canonical_block_by_slot(self, slot: int) -> BaseBeaconBlock:
+        """
+        Returns the block header with the given slot in the canonical chain.
+
+        Raises BlockNotFound if there's no block with the given slot in the
+        canonical chain.
+
+        NOTE: copied from ``eth.db.header.HeaderDB.get_canonical_block_header_by_number``,
+        and changed "header" to "block".
+        """
+        return self._get_canonical_block_by_slot(self.db, slot)
+
+    @classmethod
+    def _get_canonical_block_by_slot(
+            cls,
+            db: BaseDB,
+            slot: int) -> BaseBeaconBlock:
+        validate_block_number(slot, title="Slot")  # FIXME: Create validate_slot
+        """
+        NOTE: copied from ``eth.db.header.HeaderDB._get_canonical_block_header_by_number``,
+        and changed "header" to "block".
+        """
+        canonical_block_hash = cls._get_canonical_block_hash(db, slot)
+        return cls._get_block_by_hash(db, canonical_block_hash)
+
+    def get_canonical_head(self) -> BaseBeaconBlock:
+        """
+        Returns the current block at the head of the chain.
+        """
+        return self._get_canonical_head(self.db)
+
+    @classmethod
+    def _get_canonical_head(cls, db: BaseDB) -> BaseBeaconBlock:
+        """
+        NOTE: copied from ``eth.db.header.HeaderDB._get_canonical_head``,
+        and changed "header" to "block".
+        """
+        try:
+            canonical_head_hash = db[SchemaV1.make_canonical_head_hash_lookup_key()]
+        except KeyError:
+            raise CanonicalHeadNotFound("No canonical head set for this chain")
+        return cls._get_block_by_hash(db, canonical_head_hash)
+
+    def get_block_by_hash(self, block_hash: Hash32) -> BaseBeaconBlock:
+        return self._get_block_by_hash(self.db, block_hash)
+
+    @staticmethod
+    def _get_block_by_hash(db: BaseDB, block_hash: Hash32) -> BaseBeaconBlock:
```

Comment:
Curious what the justification/reason is for this pattern of implementing all of these as separate methods?  My initial reaction is :-1: but if there's a reason I'd like to hear it.